yb-voyager/cmd/importDataRandomFileBatchProducer.go

```diff
@@ -0,0 +1,156 @@
+	mu                                  sync.Mutex
+	sequentialFileBatchProducerFinished bool
+	producerWaitGroup                   sync.WaitGroup
+	producerCtxCancel                   context.CancelFunc
+}
+
+func NewRandomFileBatchProducer(task *ImportFileTask, state *ImportDataState, errorHandler importdata.ImportDataErrorHandler, progressReporter *ImportDataProgressReporter) (*RandomBatchProducer, error) {
+	sequentialFileBatchProducer, err := NewSequentialFileBatchProducer(task, state, errorHandler, progressReporter)
+	if err != nil {
+		return nil, fmt.Errorf("creating sequential file batch producer: %w", err)
+	}
+
+	return newRandomFileBatchProducer(sequentialFileBatchProducer, task), nil
+}
+
+// newRandomFileBatchProducer creates a RandomBatchProducer with the provided SequentialFileBatchProducer.
+// This unexported function allows tests to inject a testable sequential producer.
+func newRandomFileBatchProducer(sequentialFileBatchProducer BatchProducer, task *ImportFileTask) *RandomBatchProducer {
+	producerCtx, producerCtxCancel := context.WithCancel(context.Background())
+	rbp := &RandomBatchProducer{
+		sequentialFileBatchProducer: sequentialFileBatchProducer,
+		task:                        task,
+		sequentiallyProducedBatches: make([]*Batch, 0),
+		producerCtxCancel:           producerCtxCancel,
+	}
+	rbp.producerWaitGroup.Add(1)
+
+	go func() {
+		defer rbp.producerWaitGroup.Done()
+		err := rbp.startProducingBatches(producerCtx)
+		if err != nil {
+			utils.ErrExit("error producing batches for table: %s, file: %s, err: %w", rbp.task.TableNameTup, rbp.task.FilePath, err)
+		}
+	}()
+	return rbp
+}
+
+/*
+sequential batch producer is done producing all batches, and all batches(metadata) in memory have been consumed (in random order)
+*/
+func (rbp *RandomBatchProducer) Done() bool {
+	rbp.mu.Lock()
+	defer rbp.mu.Unlock()
+	return rbp.sequentialFileBatchProducerFinished && len(rbp.sequentiallyProducedBatches) == 0
+}
+
+// Close cleans up resources used by the RandomBatchProducer.
+func (rbp *RandomBatchProducer) Close() {
+	// stop producer goroutine
+	rbp.producerCtxCancel()
+	rbp.producerWaitGroup.Wait()
+
+	// close sequential file batch producer
+	if rbp.sequentialFileBatchProducer != nil {
+		rbp.sequentialFileBatchProducer.Close()
+	}
+}
+
+func (rbp *RandomBatchProducer) IsBatchAvailable() bool {
+	if rbp.Done() {
+		return false
+	}
+	// not done. producer may still be producing batches in parallel
+	rbp.mu.Lock()
+	defer rbp.mu.Unlock()
+	return len(rbp.sequentiallyProducedBatches) > 0
+}
+
+func (rbp *RandomBatchProducer) NextBatch() (*Batch, error) {
+	rbp.mu.Lock()
+	defer rbp.mu.Unlock()
+
+	if len(rbp.sequentiallyProducedBatches) == 0 {
+		// this could be either because
+		// 1. sequential file batch producer is done
+		// 2. sequential file batch producer is not done, and is producing in parallel.
+		// It is the responsibility of the caller to deal with both the scenarios, we do not want to block the caller.
```

Comment:
How will this be handled? In the caller, will we skip the error and continue with other files?